### PR TITLE
grindable/juiceable stacks

### LINKED ISSUE #4179: Grindable/Juiceable things don't take stack counts into account.
Which means that a grindable stack of plasma will give you the same amount of plasma for 1, 5 or 30 sheets.

### PR DESCRIPTION
**Pull request by plinyvic**
reagent grinder now takes stacks into account
added ScaleSolution that allows scaling of solution contents

## About the PR 

Fixes #4179 
**Changelog**

- fix: fixed reagent grinder giving correct output with stacked items

:cl: plinyvic

### CODE REVIEW

**Review by Zumorica — CHANGES_REQUESTED** [review]
This needs to be made more [ECS](https://hackmd.io/@ss14/ecs) instead of coupling StackComponent and JuiceableComponent like this

The way to improve this is the following:
1. Create and raise a directed event for scaling the solution of an item being juiced.
2. Subscribe to this event, directed to StackComponent. Multiply the event instance multiplier by stack count.
3. Use multiplier from raised event to scale the solution.

Now, I do think that Juiceable/Grindable need a very big refactor, but that's out of scope.

**Comment by Zumorica** [comment]
The changelog in the PR's description is broken, either fix it or remove it

**Comment by plinyvic** [comment]
should be fixed

File: Content.Server/Kitchen/EntitySystems/ReagentGrinderSystem.cs [hunk 1]
```diff
@@ -275,12 +283,17 @@ private void DoWork(ReagentGrinderComponent component, IEntity user, SharedReage
                         {
                             if (!item.HasTag("Grindable")) continue;
                             if (!item.TryGetComponent<SolutionContainerComponent>(out var solution)) continue;
-                            if (component.HeldBeaker.CurrentVolume + solution.CurrentVolume > component.HeldBeaker.MaxVolume) continue;
+                            var juiceEvent = new JuiceableScalingEvent(); // default of scalar is always 1.0
+                            if (item.HasComponent<StackComponent>())
+                            {
+                                RaiseLocalEvent<JuiceableScalingEvent>(item.Uid, juiceEvent);
+                            }
```
Comment (mirrorcult) [inline Content.Server/Kitchen/EntitySystems/ReagentGrinderSystem.cs:290]: you don't need to do the HasComponent check; the effect args just won't be modified if there isn't a stackcomp

also, make sure that you're raising the event directed by passing in false for the third arg
```suggestion
                            RaiseLocalEvent<JuiceableScalingEvent>(item.Uid, juiceEvent, false);
```

File: Content.Server/Kitchen/EntitySystems/ReagentGrinderSystem.cs [hunk 2]
```diff
@@ -40,11 +42,17 @@ public override void Initialize()
             SubscribeLocalEvent<ReagentGrinderComponent, PowerChangedEvent>((_, component, _) => EnqueueUiUpdate(component));
             SubscribeLocalEvent<ReagentGrinderComponent, InteractHandEvent>(OnInteractHand);
             SubscribeLocalEvent<ReagentGrinderComponent, InteractUsingEvent>(OnInteractUsing);
+            SubscribeLocalEvent<StackComponent, JuiceableScalingEvent>(JuiceableScaling);
+        }
+
+        private void JuiceableScaling(EntityUid uid, StackComponent component, JuiceableScalingEvent args)
+        {
+            args.Scalar = component.Count; // set scalar to amount of items in stack
```
Comment (Zumorica) [inline Content.Server/Kitchen/EntitySystems/ReagentGrinderSystem.cs:50]:
```suggestion
            args.Scalar *= component.Count; // multiply scalar by amount of items in stack
```

**Review by Zumorica — CHANGES_REQUESTED** [review]

**Pull request merged**